WT-10514 Add a new option for verify command to process all tables

mm-ng opened the pull request.

mm-ng commented:
It seems like we do use the verify operation on LSM and we do it on a bunch of our tests so I'm going to leave it in.

etienneptl commented on src/schema/schema_worker.c:
```diff
@@ -95,6 +95,13 @@ __wt_schema_worker(WT_SESSION_IMPL *session, const char *uri,
     if (skip)
         return (0);
 
+    /* FIXME-WT-10520 - Let verify process tiered storage related tables and objects once it is
```
Multiline comment format

etienneptl commented on src/utilities/util_verify.c:
```diff
@@ -31,20 +31,43 @@ usage(void)
+        ret = util_err(session, ret, "session.verify: %s", uri);
+    else {
+        /*
+         * Verbose configures a progress counter, move to the next line.
```
One-line comment (I know it's not you, but we can improve it)

etienneptl commented on src/utilities/util_verify.c:
```diff
@@ -31,20 +31,43 @@ usage(void)
+
+    if ((ret = session->verify(session, uri, config)) != 0)
+        ret = util_err(session, ret, "session.verify: %s", uri);
+    else {
```
We can probably improve this by writing `else if(verbose)` directly

mm-ng replied:
Good suggestion, thanks!

etienneptl commented on src/utilities/util_verify.c:
```diff
@@ -122,14 +136,40 @@ util_verify(WT_SESSION *session, int argc, char *argv[])
+            if (ret == ENOENT)
+                ret = 0;
+
+            ret = util_err(session, ret, "%s: WT_SESSION.open_cursor", WT_METADATA_URI);
```
Can we use `WT_ERR` here?

mm-ng replied:
Yeah, I can give this a try, for some reason I thought that wasn't supported in utility files.

keitharnoldsmith commented on src/utilities/util_verify.c:
```diff
@@ -122,14 +133,34 @@ util_verify(WT_SESSION *session, int argc, char *argv[])
+            WT_ERR(util_err(session, ret, "%s: WT_SESSION.open_cursor", WT_METADATA_URI));
+        }
+
+        while ((ret = cursor->next(cursor)) == 0) {
```
What should `util_verify()` return when run on all tables? It looks like it will always return an error value (WT_NOTFOUND?) from the last call to `cursor->next()`. This probably isn't what we want.

mm-ng replied:
Thanks for pointing that out, I didn't implement this correctly at all. After thinking about it I'd like it to still return an error if verify fails on one of the tables but I do think it should keep going and try to verify the rest of the tables.

etienneptl replied:
It could hide an actual error in the `cursor.next` implementation though but I cannot see any other way to avoid this as of now unless we know how many entries we are supposed to go through.

keitharnoldsmith replied:
> It could hide an actual error in the cursor.next implementation

You could explicitly check for the expected error -- `WT_NOTFOUND` or whatever gets returned when we hit the end of the table, and if there is some other error return it.

But the primary goal here is to verify the user's data. This isn't a correctness test for cursors. So IMO we need to worry about this case too much.

keitharnoldsmith replied:
> Thanks for pointing that out, I didn't implement this correctly at all. After thinking about it I'd like it to still return an error if verify fails on one of the tables but I do think it should keep going and try to verify the rest of the tables.

No worries. Everyone has PRs that don't work on the first go.  

I agree that it is fine to keep going and verify all the tables. You'll just have to decide whether to return the first or last error. It probably doesn't matter as long as they all get reported in the output.

It might be interesting to have an option where the user can say whether they want to stop on the 1st error. But that seems like a separate, additional ticket and not something that should happen here.

keitharnoldsmith reviewed (commented):
Monica (and @etienneptl), this is a nice idea -- It makes sense to be able to verify everything in the system with one command. It seems like a real-world usage scenario. 

I don't think we can just iterate over every entry in the metadata and verify all of them, however. Typically each WT file will have multiple metadata entries. For example `table:foo`, `colgroup:foo`, and `file:foo.wt` will all be names for the same thing. Verifying all three names means we will do 3x more work than we need to. 

There is a similar problem with the Tiered Table exclusion in `wt_schema_worker()`, as a tiered table will also have a `table:` metadata entry and a `file:` metadata entry.

This change also needs to:

- update `docs/command-line.dox` to describe the new behavior.
- add test case(s) to cover the new behavior, probably in python suite.

mm-ng commented:
> Monica (and @etienneptl), this is a nice idea -- It makes sense to be able to verify everything in the system with one command. It seems like a real-world usage scenario.
> 
> I don't think we can just iterate over every entry in the metadata and verify all of them, however. Typically each WT file will have multiple metadata entries. For example `table:foo`, `colgroup:foo`, and `file:foo.wt` will all be names for the same thing. Verifying all three names means we will do 3x more work than we need to.
> 
> There is a similar problem with the Tiered Table exclusion in `wt_schema_worker()`, as a tiered table will also have a `table:` metadata entry and a `file:` metadata entry.
> 
> This change also needs to:
> 
> * update `docs/command-line.dox` to describe the new behavior.
> * add test case(s) to cover the new behavior, probably in python suite.

Thanks for your feedback @keitharnoldsmith !

Yes, I wondered about the same thing. In that case would it be better to only run verify for items with a `table:` URI so that we don't double up with `colgroup:` and `file:` ? Something else I was wondering about is whether we expect verify to work with the history store as well? I get a `Device or resource busy` error when trying to verify it. I also noticed that the history store was also excluded from the `wt list` command so maybe it should not be included.

etienneptl commented on src/docs/command-line.dox:
```diff
@@ -469,8 +469,9 @@ The \c upgrade command has no command-specific options.
-the data source is correct, and failure if the data source is corrupted.
+The \c verify command verifies the specified table if a URI argument is provided,
+exiting success if the data source is correct, and failure if the data source is corrupted.
+If no URI argument is provided, it verifies all the tables in the database.
```
Should we specify if verify keeps going after the first failure encountered?

keitharnoldsmith commented:
>  In that case would it be better to only run verify for items with a table: URI so that we don't double up with colgroup: and file: ?

Yes, that makes sense. We'd also want to include `lsm:` items, since it doesn't look like theythey get a `table:` entry.

This will make the tiered table exclusion harder, as we'll need to differentiate a Tiered `table:` from a regular one.

mm-ng commented:
@keitharnoldsmith @etienneptl I've modified the PR so that we only verify `table:` and `lsm:` entries as per Keith's comment on the double-up of work.

With the tiered tables, I'm not sure if it's better for us to raise an error that it's an unsupported operation or if we should just silently skip over verifying the tiered tables. The way I've implemented it currently just skips over any tiered tables. The user will only see the "Operation not supported message if they run the verify command with a specific URI argument starting with `object:`, `tier` or `tiered`.

etienneptl reviewed (approved):
LGTM, thanks for doing this work!

keitharnoldsmith commented on src/docs/command-line.dox:
```diff
@@ -469,8 +469,12 @@ The \c upgrade command has no command-specific options.
-the data source is correct, and failure if the data source is corrupted.
+The \c verify command verifies the specified table if a URI argument is provided,
+exiting success if the data source is correct, and failure if the data source is corrupted.
+If no URI argument is provided, it verifies all the tables in the database. Verify will
+keep going and try to verify the rest of the tables in the scenario that it encounters an
+error when verifying a table. Currently, the verify operation is not supported with tiered
+storage and the history store file.
```
> If no URI ... error when verifying a table

I think this would be simpler as:

> If no URI argument is provided, it verifies each table in the database, reporting on any that are damaged.

Also, please change `tiered storage` to `tiered tables`.

keitharnoldsmith commented on src/schema/schema_worker.c:
```diff
@@ -124,7 +130,11 @@ __wt_schema_worker(WT_SESSION_IMPL *session, const char *uri,
+             * FIXME-WT-10520 - Don't skip verifying tiered tables once it is supported. (Change the
+             * statement below to skip = false.)
+             */
+            skip = (file_func == __wt_verify) && WT_PREFIX_MATCH(colgroup->source, "tiered:");
```
I think we should return ENOTSUP here if we have a tiered table. 

In theory we could have a complex column group that is a mix of tiered tables and other tables and we wouldn't verify those other tables. but since tiered tables aren't supported yet, I don't think we need to worry about that, and it would be better to tell the user (via ENOTSUP) that we didn't verify the tiered table.

keitharnoldsmith reviewed (approved):
LGTM.

I left a couple of small comments. I don't need to re-review those changes.

mm-ng commented:
evergreen merge

mm-ng commented:
Confirmed that the patch failure is not related to my change.
